pandas/_libs/tslibs/strptime.pyx

```diff
@@ -754,3 +775,232 @@ cdef tzinfo parse_timezone_directive(str z):
                      (microseconds // 60_000_000))
     total_minutes = -total_minutes if z.startswith("-") else total_minutes
     return pytz.FixedOffset(total_minutes)
+
+
+cdef array_strptime_object(ndarray[object] values, object fmt):
+    """
+    Calculates the datetime structs represented by the passed array of strings and
+    returns it as an object array.
+
+    values : ndarray of string-like objects
+    fmt : string-like regex
+    """
+    cdef:
+        Py_ssize_t i, n = len(values)
+        npy_datetimestruct dts
+        int64_t[:] iresult
+        object[:] result_timezone
+        int year, month, day, minute, hour, second, weekday, julian
+        int week_of_year, week_of_year_start, parse_code, ordinal
+        int iso_week, iso_year
+        int64_t us, ns
+        object val, group_key, ampm, found, timezone
+        dict found_key
+
+    dts.us = dts.ps = dts.as = 0
+
+    result = np.empty(n, dtype='O')
+    iresult = result.view('i8')
+
+    for i in range(n):
+        val = values[i]
+        if isinstance(val, str):
+            if val in nat_strings:
+                iresult[i] = NPY_NAT
+                continue
+        else:
+            if checknull_with_nat(val):
+                iresult[i] = NPY_NAT
+                continue
+            else:
+                val = str(val)
+
+        # exact matching
+        if exact:
+            found = format_regex.match(val)
+            if not found:
+                result[i] = int(val)
+            if len(val) != found.end():
+                result[i] = int(val)
+
+        # search
+        else:
+            found = format_regex.search(val)
+            if not found:
+                result[i] = int(val)
+
+        iso_year = -1
```

Comment:
we are going to want to factor this out into a function instead of repeating this for both